agent/consul/acl_test.go

```diff
@@ -346,15 +348,17 @@ func testPolicyForID(policyID string) (bool, *structs.ACLPolicy, error) {
 			RaftIndex:   structs.RaftIndex{CreateIndex: 1, ModifyIndex: 2},
 		}, nil
 	case "node-wr":
-		return true, &structs.ACLPolicy{
+		p := &structs.ACLPolicy{
 			ID:          "node-wr",
 			Name:        "node-wr",
 			Description: "node-wr",
 			Rules:       `node_prefix "" { policy = "write"}`,
 			Syntax:      acl.SyntaxCurrent,
 			Datacenters: []string{"dc1"},
 			RaftIndex:   structs.RaftIndex{CreateIndex: 1, ModifyIndex: 2},
-		}, nil
+		}
+		p.SetHash(false)
```

Comment:
I'm wondering about this fix. Why does the ACL resolver need to call `SetHash` ? Shouldn't that be done when the policy is created instead of when it is read?

Reply:
We do set the hash on creation in the PolicySet endpoint: https://github.com/hashicorp/consul/blob/main/agent/consul/acl_endpoint.go#L1204

testPolicyForID is used by a test resolver delegate, so I think it's necessary to do it on reads here because there is no write beyond the policy existing in this function. It resolves these hard coded policies as if they had been created by some other process. (But maybe I'm misunderstanding your question)

Reply:
Thanks! I think that suggests that this fix is reasonable (because it better matches a production flow), which is reassuring.

I think my broader question is about the call to `SetHash` that caused the data race. The call in `resolveWithCache` ([here](https://github.com/hashicorp/consul/blob/v1.10.0/agent/structs/acl.go#L779)).

That seems like a very strange place to be calling `SetHash`. From my reading of the code this method is called in read flows, which should already have a hash set from when the policy was written. And if not, why would `SetHash` be called here, and not in a more core place (like the state store or RPC handler) ? My thinking is that if this `SetHash` call is misplaced, it has the potentially to cause or hide bugs, either now or in the future. I guess maybe it was done defensively because we are about to read the hash, but I think sometimes this kind of defense can actually introduce other problems.

I think answering that question isn't a blocker for this PR, but it may be worth exploring further to see if we can rationalize where `SetHash` is called.

Reply:
Yeah that's a good point about `resolveWithCache`. I'm not sure why it's done there. At first I thought it could have been a new field, where policies before some version doesn't have it, but the Hash field has existed since the [New ACLs PR](https://github.com/hashicorp/consul/pull/4791).

It also doesn't seem to be used for ACL compatibility, since the function to extract policies from embedded tokens sets the hash itself:
https://github.com/hashicorp/consul/blob/0527dcff5753ddf2943581c73e1284ea81f66e3c/agent/structs/acl.go#L475

Same with these other synthetic policy extractors:
https://github.com/hashicorp/consul/blob/0527dcff5753ddf2943581c73e1284ea81f66e3c/agent/structs/acl.go#L189
https://github.com/hashicorp/consul/blob/0527dcff5753ddf2943581c73e1284ea81f66e3c/agent/structs/acl.go#L233

My sense is that it was there for tests like this one that don't set the hash in advance. I don't see a non-test use for it.

Agreed that it doesn't need to be handled in this PR.